src/session/session_dhandle.c

```diff
@@ -282,56 +282,309 @@ __wt_session_release_dhandle(WT_SESSION_IMPL *session)
     return (ret);
 }
 
+/*
+ * __session_fetch_checkpoint_meta --
+ *     Retrieve information about the selected checkpoint. Notes on the returned values are found
+ *     under __session_lookup_checkpoint.
+ */
+static int
+__session_fetch_checkpoint_meta(WT_SESSION_IMPL *session, const char *ckpt_name,
+  WT_CKPT_SNAPSHOT *info_ret, uint64_t *snapshot_time_ret, uint64_t *stable_time_ret,
+  uint64_t *oldest_time_ret)
+{
+    /* Get the timestamps. */
+    WT_RET(__wt_meta_read_checkpoint_timestamp(
+      session, ckpt_name, &info_ret->stable_ts, stable_time_ret));
+    WT_RET(
+      __wt_meta_read_checkpoint_oldest(session, ckpt_name, &info_ret->oldest_ts, oldest_time_ret));
+
+    /* Get the snapshot. */
+    WT_RET(__wt_meta_read_checkpoint_snapshot(session, ckpt_name, &info_ret->snapshot_min,
+      &info_ret->snapshot_max, &info_ret->snapshot_txns, &info_ret->snapshot_count,
+      snapshot_time_ret));
+
+    /*
+     * If we successfully read a null snapshot, set the min and max to WT_TXN_MAX so everything is
+     * visible. (Whether this is desirable isn't entirely clear, but if we leave them set to
+     * WT_TXN_NONE, then nothing is visible, and that's clearly not useful. The other choices are to
+     * fail, which doesn't help, or to signal somehow to the checkpoint cursor that it should run
+     * without a dummy transaction, which doesn't work.)
+     */
+    if (info_ret->snapshot_min == WT_TXN_NONE && info_ret->snapshot_max == WT_TXN_NONE) {
+        info_ret->snapshot_min = info_ret->snapshot_max = WT_TXN_MAX;
```

Comment:
The reasoning is that if the checkpoint has no snapshot, the best we can do is make everything in it visible; if we fail, you'll just never be able to read that checkpoint, and that's not much use. 

This should only ever happen for old named checkpoints that don't have a snapshot in the metadata. Otherwise, the snapshot info should be there and get read properly.

Reply:
Thanks, I understood the reason behind the change.

I got another doubt while going through the code. Suppose we have a named checkpoint that was taken in the previous database generation that includes some torn transactions. Opening that checkpoint after restart can cause problems as we reset the transaction ids that are part of the page, so even if we use the saved snapshot as the cursor snapshot, we may get the wrong data.

I believe this can be fixed by not advancing the btree->base_write_gen if the dhandle is for the checkpoint?

Reply:
I _think_ that works because the snapshot btree's base write generation is the one from the snapshot. Or at least that's what I believed when I was writing the code. I'm not sure that case is included in any of the current tests though and it probably should be. Will go do that; if it doesn't work I'll find out...

Reply:
It does not work. At all. Nice catch! Need to think about how to deal with it.

Reply:
Yes, that is the correct approach to fix the problem. That change will not reset any transaction ids present on the checkpoint pages irrespective of whether it is accessed even after the restart.

`run_write_gen` is the `connection->base_write_gen` recorded for each btree.
`write_gen` is the counter that is incremented for every write.

Reply:
Thanks! Am running more tests but will push it when those finish, assuming nothing else breaks.

Reply:
Pushed in [4570ff5](https://github.com/wiredtiger/wiredtiger/pull/7661/commits/4570ff593aa766c0533bf2ba6f972307bb495914).